### PR TITLE
Add AMP as a metrics destination

### PR DESCRIPTION
**Pull request by sky333999**
# Description of changes
Introduce the ability to publish metrics to Amazon Managed Prometheus as a destination.
Key highlights:
* Add `metrics_destinations` under `metrics` section of the agent config. The two supported destinations are `cloudwatch` (default if unspecified) and `amp`.
* `amp` destination requires the `workspace_id` as a required input. The remote write endpoint is built using a regex on top of the workspace ID.
* Leverages the [`prometheusremotewriteexporter`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/prometheusremotewriteexporter/README.md) to publish metrics to AMP.
* Attaches the [`sigv4authextension`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/sigv4authextension/README.md) to sign API calls made to AMP using AWS credentials.
* Extract `rollup_dimensions` and `drop_original_metrics` functionality into a new processor that can be applied to non-cloudwatch destinations.

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
Manually tested changes on an EC2 to publish `cpu` metrics to AMP.
Screenshot 2024-05-28 at 11 26 16 PM

# Requirements
_Before commit the code, please do the following steps._
1. Run `make fmt` and `make fmt-sh`
2. Run `make lint`

### CODE REVIEW

File: translator/translate/otel/exporter/prometheusremotewrite/testdata/config.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  multiplier: 1.5
+  max_interval: 30000000000
+  max_elapsed_time: 300000000000
+remote_write_queue:
+  enabled: true
+  queue_size: 10000
+  num_consumers: 5
+external_labels: []
+write_buffer_size: 524288
+endpoint: "https://aps-workspaces..amazonaws.com/workspaces/ws-12345/api/v1/remote_write"
```
Comment (jefchien) [inline translator/translate/otel/exporter/prometheusremotewrite/testdata/config.yaml:20]: Can you set the region in your unit test so this looks more like how a valid endpoint would?

File: translator/translate/otel/exporter/prometheusremotewrite/translator.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+// amp or prometheus section of the JSON config.
+func (t *translator) Translate(conf *confmap.Conf) (component.Config, error) {
+	if conf == nil || !conf.IsSet(AMPSectionKey) {
+		return nil, &common.MissingKeyError{ID: t.ID(), JsonKey: AMPSectionKey}
+	}
+	cfg := t.factory.CreateDefaultConfig().(*prometheusremotewriteexporter.Config)
+	cfg.ClientConfig.Auth = &configauth.Authentication{AuthenticatorID: component.NewID(component.MustNewType(common.SigV4Auth))}
+	cfg.ResourceToTelemetrySettings = resourcetotelemetry.Settings{Enabled: true, ClearAfterCopy: true}
+	if value, ok := common.GetString(conf, common.ConfigKey(AMPSectionKey, common.WorkspaceIDKey)); ok {
+		ampEndpoint := "https://aps-workspaces." + agent.Global_Config.Region + ".amazonaws.com/workspaces/" + value + "/api/v1/remote_write"
```
Comment (lisguo) [inline translator/translate/otel/exporter/prometheusremotewrite/translator.go:51]: Do we have any validations on `agent.Global_Config.Region`? Wondering if customers can somehow set this field to be something unexpected, triggering false prometheus calls.

File: translator/translate/otel/pipeline/host/translators.go [hunk 3]
```diff
@@ -36,11 +42,33 @@ func NewTranslators(conf *confmap.Conf, os string) (pipeline.TranslatorMap, erro
 	hasHostPipeline := hostReceivers.Len() != 0
 	hasDeltaPipeline := deltaReceivers.Len() != 0
 
-	if hasHostPipeline {
-		translators.Set(NewTranslator(common.PipelineNameHost, hostReceivers))
+	otherExporters := map[string]common.Translator[component.Config]{
+		common.AMPKey: prometheusremotewrite.NewTranslatorWithName("amp"),
+	}
+
+	if !conf.IsSet(metricsDestinationsKey) || conf.IsSet(common.ConfigKey(metricsDestinationsKey, "cloudwatch")) {
```
Comment (lisguo) [inline translator/translate/otel/pipeline/host/translators.go:49]: nit: have a const for `"cloudwatch"`

File: translator/translate/otel/pipeline/host/translators.go [hunk 4]
```diff
@@ -36,11 +42,33 @@ func NewTranslators(conf *confmap.Conf, os string) (pipeline.TranslatorMap, erro
...
+			if _, ok := common.GetString(conf, common.ConfigKey(common.MetricsKey, common.MetricsDestinationsKey, key)); ok {
+				exporters.Set(otherExporters[key])
+			}
+		}
+		if exporters.Len() != 0 {
+			// PRW exporter does not need the delta conversion.
+			receivers := common.NewTranslatorMap[component.Config]()
+			receivers.Merge(hostReceivers)
+			receivers.Merge(deltaReceivers)
+			translators.Set(NewTranslator(fmt.Sprintf("%s_other", common.PipelineNameHost), receivers, exporters))
```
Comment (lisguo) [inline translator/translate/otel/pipeline/host/translators.go:70]: What is "%s_other" for?

File: translator/translate/otel/exporter/prometheusremotewrite/testdata/config.yaml (hunk 1 above)
Comment (sky333999) [reply translator/translate/otel/exporter/prometheusremotewrite/testdata/config.yaml]: Updated

File: processor/rollupprocessor/README.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# Rollup Processor
+
+The Rollup Processor creates new data points with attribute sets that are aggregated (rolled up) from the original data points.
+For example, specifying an attribute set of `["Attr1","Attr2"]` would roll up on those two attributes, creating a new data point
+with only `"Attr1"` and `"Attr2"` and dropping all other attributes.
```
[the remark this answers is not in the document]
Comment (jefchien) [reply processor/rollupprocessor/README.md:5]: That's the way our `aggregation_dimensions` works.

> aggregation_dimensions – Optional. Specifies the dimensions that collected metrics are to be aggregated on. For example, if you roll up metrics on the AutoScalingGroupName dimension, the metrics from all instances in each Auto Scaling group are aggregated and can be viewed as a whole.
>
>You can roll up metrics along single or multiple dimensions. For example, specifying [["InstanceId"], ["InstanceType"], ["InstanceId","InstanceType"]] aggregates metrics for instance ID singly, instance type singly, and for the combination of the two dimensions.
>
>You can also specify [] to roll up all metrics into one collection, disregarding all dimensions.

File: translator/translate/otel/pipeline/host/translators.go (hunk 4 above)
Comment (jefchien) [reply translator/translate/otel/pipeline/host/translators.go:83]: It was meant for all non-cloudwatch destinations, which is an artifact from a previous change. Will make it more specific to AMP.

File: translator/translate/otel/exporter/prometheusremotewrite/translator.go (hunk 2 above)
Comment (jefchien) [reply translator/translate/otel/exporter/prometheusremotewrite/translator.go:51]: We don't. I can add that here or separately in a different PR.

File: processor/rollupprocessor/config.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+type Config struct {
+	// AttributeGroups are the groups of attribute names that will be used
+	// to create rollup data points with. The number of distinct groups will
+	// match the number of duplicate data points that are created with those
+	// attributes.
+	AttributeGroups [][]string `mapstructure:"attribute_groups,omitempty"`
+	// DropOriginal is the names of metrics where the original data points should
+	// be dropped. This is used with the AttributeGroups to reduce the number of
+	// data points sent to the exporter.
+	DropOriginal []string `mapstructure:"drop_original,omitempty"`
```
[the remark this answers is not in the document]
Comment (jefchien) [reply processor/rollupprocessor/config.go:15]: Tried to keep it similar to the JSON config, but I'm open to changing it.
> drop_original_metrics – Optional. If you are using the aggregation_dimensions field in the metrics section to roll up metrics into aggregated results, then by default the agent sends both the aggregated metrics and the original metrics that are separated for each value of the dimension. If you don't want the original metrics to be sent to CloudWatch, you can specify this parameter with a list of metrics. The metrics specified along with this parameter don't have their metrics by dimension reported to CloudWatch. Instead, only the aggregated metrics are reported. This reduces the number of metrics that the agent collects, reducing your costs.

[the remark this answers is not in the document]
Comment (lisguo) [reply processor/rollupprocessor/config.go:15]: ah okay thats prob fine

**Review by lisguo — APPROVED** [review]

File: translator/translate/metrics/config/registered_metrics.go [hunk 7]
```diff
@@ -54,9 +52,9 @@ var Registered_Metrics_Windows = map[string][]string{
...
 var DisableWinPerfCounters = map[string]bool{
-	"statsd":      true,
-	"procstat":    true,
-	"nvidia_smi":  true,
-	common.JmxKey: true,
-	"otlp":        true,
+	"statsd":     true,
+	"procstat":   true,
+	"nvidia_smi": true,
+	"jmx":        true,
```
Comment (okankoAMZ) [inline translator/translate/metrics/config/registered_metrics.go:58]: how is this related to AMP?

File: translator/translate/otel/pipeline/host/translators.go [hunk 8]
```diff
@@ -35,25 +41,36 @@ func NewTranslators(conf *confmap.Conf, os string) (pipeline.TranslatorMap, erro
 	})
 
 	switch v := conf.Get(common.ConfigKey(common.MetricsKey, common.MetricsCollectedKey, common.OtlpKey)).(type) {
-	case []interface{}:
+	case []any:
```
Comment (okankoAMZ) [inline translator/translate/otel/pipeline/host/translators.go:44]: just curious what is the difference here?

**Review by okankoAMZ — APPROVED** [review]
Looks good, left some questions

File: translator/translate/metrics/config/registered_metrics.go (hunk 7 above)
Comment (jefchien) [reply translator/translate/metrics/config/registered_metrics.go:58]: It's not.

File: translator/translate/otel/pipeline/host/translators.go (hunk 8 above)
Comment (jefchien) [reply translator/translate/otel/pipeline/host/translators.go:44]: Just a style difference.
